src/chain.h

```diff
@@ -627,10 +644,30 @@ class CDiskBlockIndex : public CBlockIndex
             }
         }
 
-        // ProgPow
-        if (this->nVersion & CBlockHeader::PROGPOW_BLOCK && nTime >= nPowTimeStampActive) {
-            READWRITE(nNonce64);
-            READWRITE(mixHash);
+        if (nTime >= nPowTimeStampActive) {
+            if (fCheckVeilDataHash) {
+                // We want to fail to accept this block. setting it NUll should fail all checks done on it
+                SetNull();
+                return;
+            }
+
+            int nPowType = (nVersion &
+                            (CBlockHeader::PROGPOW_BLOCK | CBlockHeader::RANDOMX_BLOCK | CBlockHeader::SHA256D_BLOCK));
+            switch (nPowType) {
+                case CBlockHeader::PROGPOW_BLOCK:
+                    READWRITE(nNonce64);
+                    READWRITE(mixHash);
+                    break;
+                case CBlockHeader::RANDOMX_BLOCK:
+                    break;
+                case CBlockHeader::SHA256D_BLOCK:
+                    READWRITE(nNonce64);
+                    break;
+                default:
+                    break;
```

Comment:
Should this also `SetNull(); return;` as it would imply they had either none of the bits set, or more than one?

Reply:
we can do that. however because of the consensus rules this can't actually happen. We will add the SetNull() as an extra redundancy though